Storage: add lifecycle rules helpers to bucket

tseaver opened the pull request:
@frankyn Note that the system test added in the last commit currently fails.  It is sending the POST body as

```python
{'lifecycle': {'rule': [{'action': {'type': 'Delete'},
                         'condition': {'age': 42}},
                        {'action': {'type': 'SetStorageClass'},
                         'condition': {'isLive': False,
                                       'matchesStorageClass': 'NEARLINE'},
                         'storageClass': 'COLDLINE'}]},
 'name': 'w-lifcycle-rules-1535755546278'}
```

and fails with a 400:

```python
{'error': {'code': 400,
           'errors': [{'domain': 'global',
                       'message': 'Invalid argument',
                       'reason': 'invalid'}],
           'message': 'Invalid argument'}}
```

I'd like help tracking down what is causing that error from the back-end's perspective.

tseaver commented:
@frankyn Please confirm that `storageClass` is supposed to be part of `action`:  the [API docs](https://cloud.google.com/storage/docs/json_api/v1/buckets#lifecycle) note it as an element of `action`, while the "GCS Object Lifecycle Management" one pager notes `storageClass` as peer of `action` and `condition`.

tseaver commented:
When I move `setStorageClass` inside `action`, the POSTed resource looks like:

```python
{'lifecycle': {'rule': [{'action': {'type': 'Delete'},
                         'condition': {'age': 42}},
                        {'action': {'storageClass': 'COLDLINE',
                                    'type': 'SetStorageClass'},
                         'condition': {'isLive': False,
                                       'matchesStorageClass': 'NEARLINE'}}]},
 'name': 'w-lifcycle-rules-1535767555303'}
```

and the 400 error response like this:

```python
{'error': {'code': 400,
           'errors': [{'domain': 'global',
                       'message': 'Invalid location constraint ""',
                       'reason': 'invalid'}],
           'message': 'Invalid location constraint ""'}}
```

frankyn commented:
I fixed the mistake in the document. Thanks for catching it. 

Expected request:
```
{'lifecycle': {'rule': [{'action': {'type': 'Delete'},
                         'condition': {'age': 42}},
                        {'action': {'storageClass': 'COLDLINE',
                                    'type': 'SetStorageClass'},
                         'condition': {'isLive': False,
                                       'matchesStorageClass': ['NEARLINE']}}]},
 'name': 'w-lifcycle-rules-1535767555303'}
```

The difference is matchesStorageClass expects an array of storage classes.

tseaver commented:
@frankyn I've moved 'setStorageClass` inside `action`, and updated `matches_storage_class` to document that it is a list, and pass lists everywhere.  However, the system test still fails with:

```python
(Pdb) pp properties
{'lifecycle': {'rule': [{'action': {'type': 'Delete'},
                         'condition': {'age': 42}},
                        {'action': {'storageClass': 'COLDLINE',
                                    'type': 'SetStorageClass'},
                         'condition': {'isLive': False,
                                       'matchesStorageClass': ['NEARLINE']}}]},
 'name': 'w-lifcycle-rules-1536091244549'}
```

```python
(pdb) pp response.json()
{'error': {'code': 400,
           'errors': [{'domain': 'global',
                       'message': 'Invalid location constraint ""',
                       'reason': 'invalid'}],
           'message': 'Invalid location constraint ""'}}
```

frankyn commented:
face palm, I was able to repro. This seems like a bug, but not yet sure. The part I missed was including the lifecycle policy on a bucket create operation.

frankyn commented:
By including location for the bucket `"location": "us"` it was able to create using the provided lifecycle policy. Will follow-up internally, it might be WAI, but if lifecycle isn't included location isn't required (also documented in ref guide: ["Defaults to US"](https://cloud.google.com/storage/docs/json_api/v1/buckets/insert).

tseaver commented:
@frankyn OK, fingers crossed that a534067 gets the system test passing.

tseaver commented on storage/google/cloud/storage/bucket.py:
```diff
@@ -99,6 +100,170 @@ def _item_to_notification(iterator, item):
+        return instance
+
+
+class LifecycleRuleSetItemStorageClass(dict):
```
4459a93

tseaver commented on storage/google/cloud/storage/bucket.py:
```diff
@@ -99,6 +100,170 @@ def _item_to_notification(iterator, item):
+        return self.get('numNewerVersions')
+
+
+class LifecycleRuleDeleteItem(dict):
```
4459a93

tseaver commented:
@frankyn 1e04aa6 adds the two bucket helper methods you suggested (`add_lifecycle_rule_delete` and ``add_lifecycle_rule_set_storage_class`), along with `clear_lifecycle_rules` (which seems best for parity / clarity).

frankyn commented:
(w.r.t to tests) The reason to test failure was to test behavior of client, but I'm open to removing this expectation. It seems unnecessary at this point.  

(w.r.t to the surface) The surface LGTM just have a question.

I'm curious about the lifecycle of `bucket.lifecycle_rules()` with the update:
Are the following two ways a user can now interact with rules (keeping backward compat)?
```
# new
bucket = storage_client.get_bucket("bucket-name")
for rule in bucket.lifecycle_rules():
  print(rule.action)
  print(rule.conditions.age)
```
OR
```
#existing
bucket = storage_client.get_bucket("bucket-name")
for rule in bucket.lifecycle_rules():
  print(rule["action"]
  print(rule["conditions"]["age"])
```

tseaver commented:
@frankyn Both [LifecycleRuleDelete](https://github.com/GoogleCloudPlatform/google-cloud-python/pull/5877/files#diff-2fcb3e6cad3a5f40748b2c3d14bc6066R200) and [LifecycleRuleSetStorageClass](https://github.com/GoogleCloudPlatform/google-cloud-python/pull/5877/files#diff-2fcb3e6cad3a5f40748b2c3d14bc6066R231) derived from `dict`, so the "existing" example works (well, the key in the dict is `condition`, not `conditions`, but that is what the API expects).

Neither one defines an `action` or a `conditions` property / attribute.

frankyn commented on storage/google/cloud/storage/bucket.py:
```diff
@@ -943,8 +1115,46 @@ def lifecycle_rules(self, rules):
+        self.lifecycle_rules = []
+
+    def add_lifecycle_delete_rule(self, **kw):
+        """Add a "delete" rule to lifestyle rules configured for this bucket.
```
Could you add an example snippet in the documentation:

```python
from google.cloud import storage

storage_client = storage.Client()
bucket = storage_client.get_bucket("bucket-name")
bucket.add_lifecycle_delete_rule(age=2)
bucket.patch()
```

tseaver replied:
d7674b1

frankyn commented on storage/google/cloud/storage/bucket.py:
```diff
@@ -943,8 +1115,46 @@ def lifecycle_rules(self, rules):
+
+    def add_lifecycle_set_storage_class_rule(self, storage_class, **kw):
+        """Add a "delete" rule to lifestyle rules configured for this bucket.
+
```
Could you add an example snippet in the documentation:

```python
from google.cloud import storage

storage_client = storage.Client()
bucket = storage_client.get_bucket("bucket-name")
bucket. add_lifecycle_set_storage_class_rule(storage_class="COLDLINE", 
    matches_storage_class=["NEARLINE"])
bucket.patch()
```

tseaver replied:
d7674b1

tseaver commented on storage/google/cloud/storage/bucket.py:
```diff
@@ -99,6 +100,170 @@ def _item_to_notification(iterator, item):
+        return instance
+
+    @property
+    def age(self):
```
The 'created_before' one does a conversion to a Python date instance.

tseaver replied:
I can add `rule.action` and `rule.conditions` pretty easily (I have them stashed against this branch, actually).

tseaver commented:
Merging after an "LGTM" from @frankyn in chat.

frankyn commented:
Thanks Tres!
